### ISSUE TITLE
self.log_dict() doesn't support dictionary of Metric objects

### ISSUE THREAD

**Issue by itsikad** [PyTorchLightning/pytorch-lightning#4255]
## 🐛 Bug

Calling self.log_dict with a dictionary of Metric object(s) results in 'Trying to pass too many CPU scalars to CUDA kernel!' error. Unlike calling self.log(metric) directly, in this case the internal states are not transferred to device.

### Please reproduce using

https://colab.research.google.com/drive/1a7pjASi5_TiSc_XqEFB2Rfd4yaIcFqe2?usp=sharing

### Expected behavior

The self.log_dict() method used to work properly with a dictionary of Train/EvalResults objects and is a great solution for tracking multiple metrics in training/validation.

### Environment

 - PyTorch Version (e.g., 1.0): 1.6.0
 - Pytroch-lightning: 1.0.2
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.6.9
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: 2 x 2080Ti
 - Any other relevant information: distributed backend: dp, ddp

**Comment by pbmstrk** [comment]
From the stacktrace it seems the error occurs not when `log_dict()` is called but when calculating the metrics in the lines

```
 # Update metrics
 for k,m in self.train_metrics.items():
     m(output, torch.ones_like(output))
```

When running the on the CPU no errors occur, so this is caused because tensors initialized within the metric class are not on the correct device.

Your code works when replacing,

```
self.layer = torch.nn.Linear(32, 2)
self.train_metrics = {'mse': MeanSquaredError(), 'mae': MeanAbsoluteError()}
```

with

```
self.mse = MeanSquaredError()
self.mae = MeanAbsoluteError()
self.train_metrics = {'mse': self.mse, 'mae': self.mae}
```

I believe that by doing this you ensure that the metric classes (which are themselves instances of `nn.Module`) are moved to the GPU when the model is.

**Comment by itsikad** [comment]
@pbmstrk thanks! indeed solved the issue.

I log 6 metrics during training and validation , any best practices/ideas on how to pack multiple metrics (~6-7) into a single object that update all at once when called? I tried to subclass nn.Module and add a ModuleDict. It does overcome the initial issue but now `self.dict_log()` cannot iterate over the metrics (since nn.Module has no items attribute).

**Comment by pbmstrk** [comment]
I'm not sure about best practices, but here is something simple that seems to work. Building on the idea of subclassing `nn.Module` and using `nn.ModuleDict` you can implement the `forward()` method to create a dictionary by iterating through the metrics. For example,

```
class MyMetricClass(nn.Module):
    def __init__(self, metric_dict):
        super().__init__()
        self.metrics = nn.ModuleDict(metric_dict)
        
    def forward(self, preds, targets):
        return {name: metric(preds, targets) for name, metric in self.metrics.items()}
```
 
where `metric_dict` is simply a dict of the desired metrics, e.g.

```
metric_dict = {'mse': MeanSquaredError(), 'mae': MeanAbsoluteError()}
```

Then in the `__init__()` method you can create an instance,

```
self.metric_dict = {'mse': MeanSquaredError(), 'mae': MeanAbsoluteError()}
self.metric_module = MyMetricClass(self.metric_dict)
```

and to log you can simply call the module directly in the `log_dict()` method (since it returns a `dict`), e.g.

```
# using your example
self.log_dict(self.metric_module(output, torch.ones_like(output)), on_step=False, on_epoch=True)
```

I have not tested this extensively, so it will probably break in some cases but it works in your example.

**Comment by itsikad** [comment]
Thanks again.

I took a similar approach, I implemented a `nn.Module` and 'stacked multiple metrics into `nn.ModuleDict' but I didn't return a dict (I'm not interested in computing the metrics every step) but it doesn't seem to work so far.
I'll give your suggestion a try.

**Comment by itsikad** [comment]
So I tried the most basic usage (use MeanSquaredError and self.log, same as in the docs) and got an error: 

> Only Tensors created explicitly by the user (graph leaves) support the deepcopy protocol at the moment

Reproduce (second block): https://colab.research.google.com/drive/1a7pjASi5_TiSc_XqEFB2Rfd4yaIcFqe2?usp=sharing#scrollTo=kBpMKYNU3QPy

**Comment by pbmstrk** [comment]
@itsikad i noticed this as well, it seems to be a problem with the regression metrics (for example if you try with Accuracy, the error does not occur). I opened a separate issue for this #4266

**Comment by itsikad** [comment]
Closing this issue then. Thanks for your help and time.

**Comment by jonghwanhyeon** [comment]
It seems enough to just wrap `self.train_metrics` dict using `torch.nn.ModuleDict`.

```python
self.train_metrics = torch.nn.ModuleDict({
    'mse': MeanSquaredError(), 
    'mae': MeanAbsoluteError()})
```
